Only return viewable tasks from tasks endpoint

joelclimbsthings opened the pull request:
Fixes #7647 

This PR results in the `get_tasks` endpoint, by default, only returning visible tasks. These are filtered out on snooze/dismissed status, as well as the `canView` property associated with the individual task. 

~~Also added optional `only_visible` parameter to the endpoint that defaults to `true`. That way future use-cases that demand access to snoozed/dismissed tasks wouldn't require changes to the endpoint.~~

To do:
~~- Potentially update data-store to support `is_visible` parameter, although perhaps not urgent as we don't currently have a use-case.~~ 

### Detailed test instructions:

-   Checkout branch
-   Navigate to Homescreen.
- Ensure only tasks that should be visible are (ie if you don't have WCPay installed, the WCPay task shouldn't not be shown)

joelclimbsthings committed:
Tentatively filtering based just on isVisible flag

joelclimbsthings committed:
Implementing optional parameter "only_visible" that defaults to true for get_tasks endpoint

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -114,6 +114,12 @@ public function register_routes() {
 			'/' . $this->rest_base,
 			array(
+				'args'   => array(
+					'only_visible' => array(
```
What do you think about `filter_visible` for this name?

joelclimbsthings replied:
I'm open, although `filter_visible` to me is a little ambiguous. It could mean filter _out_ visible tasks, or filter for _only_ visible tasks. I'm up for changing it, but `only_visible` feels clearer to me between the two.

joshuatf commented on client/tasks/task-list.tsx:
```diff
@@ -49,16 +49,9 @@ export const TaskList: React.FC< TaskListProps > = ( {
 	const eventPrefix = getEventPrefix( id );
 	const prevQueryRef = useRef( query );
-	const nowTimestamp = Date.now();
-	const visibleTasks = tasks.filter(
```
Looks much cleaner client-side!  I'm still a bit concerned that this doesn't address the original issue:

>Note that certain items like "dismiss" and "snooze" still need to optimistically update and potentially undo that update, which makes me unsure whether or not we can filter visibility on the REST side. cc @joelclimbsthings

Dismissing or snoozing tasks won't work as expected.

Filtering them out entirely may be problematic too if we undo snoozing or undo dismissal.

joelclimbsthings replied:
Thanks @joshuatf , this is true and I've refactored the changes to leave the logic around snooze and dismissal on the client-side. I think ultimately that's the simpler and more flexible approach. The API is, however, filtering out tasks that are `canView: false` since those are never visible on the front-end, so it seems safe to take them out.

joelclimbsthings committed:
Modifications to still handle snooze and dismiss states on client-side

joelclimbsthings commented:
Thanks for the review @joshuatf . I've addressed your comments, so let me know what you think.

I've also changed the API parameter to `only_viewable`, and the language around this to _viewable_ instead of _visible_. I'm open to changing this if better ideas surface!

Right now this is based on my `fix/task-api-snooze-dismiss` branch so the diff is a bit messy. Feel free to ignore this until I get it rebased when one of our PRs is merged.

joshuatf commented on src/Features/OnboardingTasks/Task.php:
```diff
@@ -283,7 +274,7 @@ public function get_json() {
 			'actionUrl'     => $this->action_url,
 			'isComplete'    => $this->is_complete,
-			'isVisible'     => $this->is_visible(),
+			'canView'       => $this->can_view,
```
Nice 👍

joshuatf commented on src/API/OnboardingTasks.php:
```diff
@@ -114,6 +114,12 @@ public function register_routes() {
 			'/' . $this->rest_base,
 			array(
+				'args'   => array(
+					'only_viewable' => array(
```
I like "viewable" but I'm not a huge fan of the `only_` prefix.  Can we prefix with a verb?  `filter_` possibly?

joelclimbsthings replied:
I mentioned in [this other earlier comment](https://github.com/woocommerce/woocommerce-admin/pull/7718/files#r718034370), I'm good with changing this, but `filter_viewable` seems ambiguous to me. Would `filter_only_viewable` work for you?

joelclimbsthings replied:
This seems to be moot now that I removed the endpoint parameter ✅

joshuatf commented on src/Features/OnboardingTasks/TaskList.php:
```diff
@@ -112,12 +112,27 @@ public function add_task( $args ) {
 	 * @return array
 	 */
-	public function get_json() {
+	public function get_json( $only_viewable = false ) {
```
Passing arguments to filter tasks feels a bit like an anti-pattern.  

Maybe we could create a method to allow filtering of tasks and keep the `get_json()` method simple:

```
$task_list->filter_tasks( array( 'can_view' => true ) );
$task_list->get_json()
```

joshuatf replied:
Thinking about this more, one other approach if we want to keep things simple would be to create a `get_viewable_tasks()` method on `TaskList`.  I'd be okay if `get_json()` always used `get_viewable_tasks()` given it's use is for client-side JSON.

joelclimbsthings replied:
I had actually already created a `get_viewable_tasks()` method (just above this comment), but I removed the parameter for `get_json` in a53f441620ce1dc4c788970411fe9512255c9ade. 

The only downside here is that it removes the ability to show non-viewable tasks from the endpoint at all, but may not be a need going forward so we can keep it simple.

joshuatf reviewed (commented):
Nice job, @joelclimbsthings!  Left a few structural comments, but I think this will work well with snoozing and dismissal now.

joelclimbsthings committed:
Get json on tasklist will always return viewable tasks

joelclimbsthings committed:
Fixing tests

joelclimbsthings commented:
Thanks @joshuatf , went ahead and took your suggestion to make `get_json()` always return viewable tasks for now. That removed the need for the parameter, just let me know if you have additional thoughts.

joshuatf commented on src/Features/OnboardingTasks/TaskList.php:
```diff
@@ -112,6 +112,20 @@ public function add_task( $args ) {
 	}
 
+	/**
+	 * Get only visible tasks in list.
```
Just need a `@return` value here.

joshuatf reviewed (approved):
Looks great and is testing well!  Pre-approved pending one doc block comment.  Thanks, @joelclimbsthings!

>The only downside here is that it removes the ability to show non-viewable tasks from the endpoint at all, but may not be a need going forward so we can keep it simple.

Agreed.  I can't think of a use-case right now where we'd need them, but if we do we can consider chaining some filters and extending the `TaskList` api.

joelclimbsthings committed:
Adding doc comment return value
